### PR TITLE
Add SerializeSILPass to MandatoryOpt pipeline if requested. 

### PR DESCRIPTION
**Pull request by bgogul**
This adds a SerializeSILPass just before differentiation so that we can serialize the module from "lldb -r" to support cross-cell differentiation in colab/juptyer.  (See https://github.com/apple/swift-lldb/pull/1268). This is needed to fix [SR-9805](https://bugs.swift.org/browse/SR-9805).

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/PassManager/Passes.cpp [hunk 1]
```diff
@@ -147,6 +147,22 @@ void swift::runSILTFPartitionPass(SILModule &Module) {
...
+
+  SILPassManager PM(&Module, "TensorFlow", /*isMandatoryPipeline=*/ true);
+  PM.executePassPipelinePlan(
+      SILPassPipelinePlan::getMandatoryOptPreDiffPassPipeline(
+          Module.getOptions()));
+
+  // Verify the module, if required.
+  if (Module.getOptions().VerifyAll)
+    Module.verify();
+
```
Comment (dan-zheng) [inline lib/SILOptimizer/PassManager/Passes.cpp]: minor: extra newline

**Review by dan-zheng — COMMENTED** [review]
Is the underlying problem that generated primal/adjoint/jvp/vjp functions don't have the right "qualified ownership"-ness?
If so, changing that and doing raw stage serialization seems like a more robust fix. It'd look like:
```diff
   vjp = fb.createFunction(linkage, vjpName, vjpType, vjpGenericEnv,
                           original->getLocation(), original->isBare(),
                           IsNotTransparent, original->isSerialized(),
                           original->isDynamicallyReplaceable());
-  vjp->setUnqualifiedOwnership();
+  if (!original->hasQualifiedOwnership())
+    vjp->setUnqualifiedOwnership();
// Same for primal/adjoint/jvp.
```

**Review by marcrasi — CHANGES_REQUESTED** [review]
I have an idea for a less-invasive change that will be easier to keep in sync with upstream. Not totally sure if it will work.

Three pieces:
1. Add a `SerializeForDifferentiation` to `class SILOptions`.
2. In `addMandatoryOptPipeline`, add `if (Options.SerializeForDifferentiation) { P.addSerializeSILPass() }`
3. I think that the SerializeSILPass works by calling a serialization callback that has been registered on the SILModule. So you'll have to refactor the LLDB code to do the serialization in a closure, and then register that closure with `sil_module->setSerializeSILAction()`.

**Comment by marcrasi** [comment]
I like this overall approach of running all the pre-differentiation passes before serializing in LLDB, because it pretty much guarantees that when LLDB tries to load a function and differentiate it, the differentiation pass will see the exact same SIL in the body that it would have seen if it had decided to differentiate it in the first place. This will make it a lot more likely to always work, even if there are even more problems with differentiating raw SIL that we don't know about yet.

If you add a `SerializeForDifferentiation` option, that would be a nice place to put a comment explaining this reasoning.

**Comment by bgogul** [comment]
> I like this overall approach of running all the pre-differentiation passes before serializing in LLDB, because it pretty much guarantees that when LLDB tries to load a function and differentiate it, the differentiation pass will see the exact same SIL in the body that it would have seen if it had decided to differentiate it in the first place. This will make it a lot more likely to always work, even if there are even more problems with differentiating raw SIL that we don't know about yet.
> 
> If you add a `SerializeForDifferentiation` option, that would be a nice place to put a comment explaining this reasoning.

Nice. I did think about adding an option to SILOptions and may be messing around with the mandatory pipeline. However, your suggestion of adding a call-back for serialization seems much cleaner and much less invasive.

**Comment by bgogul** [comment]
This is much cleaner now. Thanks, Marc. PTAL.

**Review by marcrasi — APPROVED** [review]
Yay!

**Review by rxwei — APPROVED** [review]

**Pull request merged**